File: sdk/ts/src/voice/pushToTalk.ts
// pushToTalk mirrors sdk/go/voice/pushtotalk.go. It opens a streaming
// transcription session, pumps audio bytes from a ReadableStream as
// AiTranscribeStreamChunk envelopes, surfaces deltas via onPartial,
// and resolves with the final transcript on completion.
//
// The SDK owns the wire protocol; the caller owns the audio source --
// terminal apps wire a Web Audio buffer, browsers wire a MediaRecorder
// (helpers like `recorderToReadableStream` may live alongside this
// file when the first browser consumer needs them).

import type { Dispatcher } from "../client/dispatcher.js";
import { newShortId } from "../client/id.js";
import {
  readServerPayload,
  type ServerMessage,
  type AiTranscribeStreamDeltaPayload,
} from "../client/wire.js";

export interface AudioFormat {
  encoding: "pcm16" | "opus" | "webm" | "wav" | (string & {});
  sampleRate: number;
  channels: number;
}

export interface PartialTranscript {
  text: string;
  isFinal: boolean;
  confidence: number;
}

export interface FinalTranscript {
  text: string;
  durationMs: number;
  provider: string;
}

export interface PushToTalkOptions {
  audio: AudioFormat;
  // ISO-639-1 language hint, e.g. "en". Empty defers to the provider.
  language?: string;
  // Provider override (e.g. "openai-realtime", "openai-whisper"). Empty
  // uses the cluster default.
  provider?: string;
  // Called on every delta. Safe to be undefined.
  onPartial?: (p: PartialTranscript) => void;
  // Cancel the session mid-stream.
  signal?: AbortSignal;
  // For consumers that prefer pull-based delta streaming over the
  // onPartial callback. Resolves to an AsyncIterable that yields
  // PartialTranscripts and ends when Complete lands or the session
  // aborts. When set, onPartial is still invoked first (both work
  // together).
  yieldDeltas?: boolean;
}

// pushToTalk resolves with the FinalTranscript when the server emits
// Complete. Throws on session-level failure or AbortSignal cancellation.
export async function pushToTalk(
  dispatcher: Dispatcher,
  audio: ReadableStream<Uint8Array>,
  opts: PushToTalkOptions,
): Promise<FinalTranscript> {
  if (!dispatcher) throw new Error("pushToTalk: dispatcher is required");
  if (!audio) throw new Error("pushToTalk: audio stream is required");
  if (!opts.audio?.encoding || !opts.audio.sampleRate || !opts.audio.channels) {
    throw new Error(
      "pushToTalk: audio format must specify encoding + sampleRate + channels",
    );
  }
  const requestId = newShortId();
  const session = new TranscriptionSession(dispatcher, requestId, opts);
  return session.run(audio);
}

class TranscriptionSession {
  private readonly dispatcher: Dispatcher;
  private readonly requestId: string;
  private readonly opts: PushToTalkOptions;

  constructor(dispatcher: Dispatcher, requestId: string, opts: PushToTalkOptions) {
    this.dispatcher = dispatcher;
    this.requestId = requestId;
    this.opts = opts;
  }

  async run(audio: ReadableStream<Uint8Array>): Promise<FinalTranscript> {
    let resolveFinal: (t: FinalTranscript) => void;
    let rejectFinal: (err: Error) => void;
    const finalPromise = new Promise<FinalTranscript>((resolve, reject) => {
      resolveFinal = resolve;
      rejectFinal = reject;
    });

    const unregister = this.dispatcher.registerStream(this.requestId, (msg: ServerMessage) => {
      const payload = readServerPayload(msg);
      if (payload?.kind === "aiTranscribeStreamDelta") {
        if (this.opts.onPartial) {
          this.opts.onPartial(deltaToPartial(payload.value));
        }
      } else if (payload?.kind === "aiTranscribeStreamComplete") {
        resolveFinal({
          text: payload.value.text ?? "",
          durationMs: numberFromWire(payload.value.durationMs),
          provider: payload.value.provider ?? "",
        });
      }
    });

    const abortListener = () => {
      // Best-effort abort -- the dispatcher may be torn down already.
      try {
        this.sendEnd(true);
      } catch {
        /* ignore */
      }
      rejectFinal(new Error("pushToTalk: aborted"));
    };
    if (this.opts.signal) {
      if (this.opts.signal.aborted) {
        unregister();
        throw new Error("pushToTalk: aborted before start");
      }
      this.opts.signal.addEventListener("abort", abortListener, { once: true });
    }

    try {
      this.sendStart();
      // Drive the audio pump in parallel with the reply listener.
      const pumpPromise = this.pumpAudio(audio);
      pumpPromise.catch((err) => {
        // Audio pump failed -- best-effort abort the session and
        // surface the error.
        try {
          this.sendEnd(true);
        } catch {
          /* ignore */
        }
        rejectFinal(err instanceof Error ? err : new Error(String(err)));
      });
      return await finalPromise;
    } finally {
      unregister();
      if (this.opts.signal) {
        this.opts.signal.removeEventListener("abort", abortListener);
      }
    }
  }

  private sendStart(): void {
    this.dispatcher.send({
      aiTranscribeStreamStart: {
        requestId: this.requestId,
        format: this.opts.audio.encoding,
        sampleRate: this.opts.audio.sampleRate,
        channels: this.opts.audio.channels,
        ...(this.opts.language ? { languageHint: this.opts.language } : {}),
        ...(this.opts.provider ? { provider: this.opts.provider } : {}),
      },
    });
  }

  private sendChunk(audio: Uint8Array): void {
    this.dispatcher.send({
      aiTranscribeStreamChunk: {
        requestId: this.requestId,
        audio: bytesToBase64(audio),
      },
    });
  }

  private sendEnd(cancel: boolean): void {
    this.dispatcher.send({
      aiTranscribeStreamEnd: {
        requestId: this.requestId,
        ...(cancel ? { cancel: true } : {}),
      },
    });
  }

  private async pumpAudio(audio: ReadableStream<Uint8Array>): Promise<void> {
    const reader = audio.getReader();
    try {
      for (;;) {
        const { value, done } = await reader.read();
        if (done) {
          this.sendEnd(false);
          return;
        }
        if (value && value.length > 0) this.sendChunk(value);
      }
    } finally {
      try {
        reader.releaseLock();
      } catch {
        /* ignore */
      }
    }
  }
}

function deltaToPartial(d: AiTranscribeStreamDeltaPayload): PartialTranscript {
  return {
    text: d.text ?? "",
    isFinal: d.isFinal === true,
    confidence: typeof d.confidence === "number" ? d.confidence : 0,
  };
}

function numberFromWire(v: string | number | undefined): number {
  if (typeof v === "number") return v;
  if (typeof v === "string" && v) {
    const n = Number(v);
    return Number.isFinite(n) ? n : 0;
  }
  return 0;
}

function bytesToBase64(bytes: Uint8Array): string {
  // Prefer Node's Buffer when present (no @types/node dependency --
  // probe via globalThis so a browser bundle doesn't trip on the
  // bare `Buffer` identifier). Fall back to btoa for browsers.
  const g = globalThis as unknown as {
    Buffer?: { from(bytes: Uint8Array): { toString(enc: string): string } };
    btoa?: (s: string) => string;
  };
  if (g.Buffer) return g.Buffer.from(bytes).toString("base64");
  if (g.btoa) {
    let binary = "";
    for (const b of bytes) binary += String.fromCharCode(b);
    return g.btoa(binary);
  }
  throw new Error("pushToTalk: no base64 encoder available (need Buffer or btoa)");
}

File: sdk/go/voice/pushtotalk.go
// Package voice is the SDK's voice subpackage. It exposes higher-
// level voice operations built on top of the wire-level
// AiTranscribeStream* protocol family carried by MemqlService.Stream.
//
// PushToTalk is the canonical entry point: open a streaming session,
// feed audio bytes from a caller-supplied reader, surface partial
// transcripts as they arrive, and resolve a final transcript when the
// session closes. The caller owns the audio source (terminal apps
// typically use malgo / miniaudio; browsers use MediaStream); the SDK
// owns the protocol.
//
// The SDK does NOT itself write the resulting utterance row into the
// space. The contract: the caller receives the final transcript via
// FinalTranscript.Text, and is responsible for invoking
// mutationSendTextUtterance (or any other downstream side-effect) on
// the active space. This keeps PushToTalk a pure transcription
// primitive that's reusable for non-chat callers (note-taking,
// command-line dictation, etc.).
package voice

import (
	"context"
	"fmt"
	"io"

	memqlv1 "github.com/znasllc-io/memql/component/grpc/gen"
	"github.com/znasllc-io/memql/core/id"
	"github.com/znasllc-io/memql/sdk/go/client"
)

// AudioFormat describes the encoding of the audio bytes being streamed.
type AudioFormat struct {
	Encoding   string // "pcm16", "opus", "webm", "wav"
	SampleRate int32  // typically 16000
	Channels   int32  // 1 (mono) or 2 (stereo)
}

// PartialTranscript fires every time the server emits a delta.
// Text is the FULL accumulated transcript so far, not an incremental
// token -- the caller can render it directly as the user types.
type PartialTranscript struct {
	Text       string
	IsFinal    bool
	Confidence float32
}

// FinalTranscript is the resolved result of a transcription session.
type FinalTranscript struct {
	Text       string
	DurationMs int64
	Provider   string
}

// Options configures a PushToTalk session.
type Options struct {
	// Audio describes the format of the bytes the caller will feed
	// into the audio Reader.
	Audio AudioFormat

	// Language is an optional ISO-639-1 hint (e.g. "en") to bias
	// transcription toward a specific language. Empty defers to the
	// provider's default.
	Language string

	// Provider optionally overrides the cluster's default transcription
	// provider (e.g. "openai-realtime", "openai-whisper").
	// Empty uses the cluster default.
	Provider string

	// ChunkBytes is the size of each AiTranscribeStreamChunk payload
	// read from the audio source. Smaller chunks = lower latency,
	// more gRPC frames. Recommended: 1280 bytes (40ms of 16kHz mono
	// PCM16). Defaults to 1280 when zero.
	ChunkBytes int

	// OnPartial is invoked from the SDK's reader goroutine on every
	// delta. Safe to be nil (caller doesn't care about partials).
	// Must not block significantly -- a slow OnPartial back-pressures
	// the stream listener.
	OnPartial func(PartialTranscript)
}

// PushToTalk runs a streaming transcription session end-to-end:
// opens the gRPC stream session, copies audio from the supplied
// io.Reader to the server in ChunkBytes-sized AiTranscribeStreamChunk
// messages, calls opts.OnPartial as deltas arrive, and returns the
// resolved FinalTranscript on completion.
//
// Lifecycle:
//   - The reader is consumed until EOF (clean end-of-utterance) or
//     a non-EOF read error (session aborted with AiTranscribeStreamEnd
//     cancel=true).
//   - On ctx cancellation, the session is aborted similarly.
//   - On stream-level failure (server crash, network drop), an error
//     is returned and no Complete event arrives.
//
// Concurrency: PushToTalk registers a stream listener with the
// dispatcher keyed on a fresh request_id. Multiple concurrent
// PushToTalk calls against the same dispatcher are safe -- each gets
// its own request_id and its own listener slot.
func PushToTalk(
	ctx context.Context,
	dispatcher *client.Dispatcher,
	audio io.Reader,
	opts Options,
) (*FinalTranscript, error) {
	if dispatcher == nil {
		return nil, fmt.Errorf("voice.PushToTalk: dispatcher is nil")
	}
	if audio == nil {
		return nil, fmt.Errorf("voice.PushToTalk: audio reader is nil")
	}
	if opts.Audio.Encoding == "" || opts.Audio.SampleRate == 0 || opts.Audio.Channels == 0 {
		return nil, fmt.Errorf("voice.PushToTalk: Options.Audio must specify Encoding + SampleRate + Channels")
	}
	chunkBytes := opts.ChunkBytes
	if chunkBytes <= 0 {
		chunkBytes = 1280
	}

	requestId := id.NewShortId()
	replies, unregister := dispatcher.RegisterStream(requestId)
	defer unregister()

	// Send the session-start envelope. Server picks up the request_id
	// here and routes every Delta/Complete back to it.
	start := &memqlv1.MemqlClientMessage{
		Payload: &memqlv1.MemqlClientMessage_AiTranscribeStreamStart{
			AiTranscribeStreamStart: &memqlv1.AiTranscribeStreamStart{
				RequestId:    requestId,
				Format:       opts.Audio.Encoding,
				SampleRate:   opts.Audio.SampleRate,
				Channels:     opts.Audio.Channels,
				LanguageHint: opts.Language,
				Provider:     opts.Provider,
			},
		},
	}
	if _, err := dispatcher.Send(start); err != nil {
		return nil, fmt.Errorf("voice.PushToTalk: send Start: %w", err)
	}

	// Stream audio in a goroutine so we can concurrently read replies.
	// readErr carries the eventual outcome of the audio-pump goroutine
	// so we can decide whether to send End{cancel:false} (clean EOF)
	// or End{cancel:true} (read error / ctx cancel).
	readErrCh := make(chan error, 1)
	go func() {
		readErrCh <- streamAudio(ctx, dispatcher, requestId, audio, chunkBytes)
	}()

	// Drain replies until Complete or stream death.
	var final *FinalTranscript
	for {
		select {
		case <-ctx.Done():
			// Best-effort abort. Don't block on Send failure -- the
			// dispatcher is likely already torn down.
			_ = sendEnd(dispatcher, requestId, true)
			<-readErrCh
			return nil, ctx.Err()

		case <-dispatcher.Done():
			return nil, fmt.Errorf("voice.PushToTalk: dispatcher closed before Complete")

		case readErr := <-readErrCh:
			// Audio pump finished. Drain Start completion path:
			// either clean EOF (we already sent End{cancel:false}) or
			// read error / ctx cancel handled by sendEnd cancel=true.
			if readErr != nil && readErr != io.EOF {
				return nil, fmt.Errorf("voice.PushToTalk: audio read: %w", readErr)
			}
			// Fall through to keep draining replies until Complete.
			readErrCh = nil

		case msg, ok := <-replies:
			if !ok {
				return nil, fmt.Errorf("voice.PushToTalk: reply channel closed before Complete")
			}
			switch p := msg.Payload.(type) {
			case *memqlv1.MemqlServerMessage_AiTranscribeStreamDelta:
				if opts.OnPartial != nil {
					opts.OnPartial(PartialTranscript{
						Text:       p.AiTranscribeStreamDelta.GetText(),
						IsFinal:    p.AiTranscribeStreamDelta.GetIsFinal(),
						Confidence: p.AiTranscribeStreamDelta.GetConfidence(),
					})
				}
			case *memqlv1.MemqlServerMessage_AiTranscribeStreamComplete:
				final = &FinalTranscript{
					Text:       p.AiTranscribeStreamComplete.GetText(),
					DurationMs: p.AiTranscribeStreamComplete.GetDurationMs(),
					Provider:   p.AiTranscribeStreamComplete.GetProvider(),
				}
				return final, nil
			}
		}
	}
}

// streamAudio copies the audio reader to the server in ChunkBytes-
// sized Chunk messages, then sends End{cancel:false} on clean EOF.
// On ctx cancel or non-EOF read error, sends End{cancel:true}.
//
// Returns nil on clean EOF, ctx.Err() on cancellation, or the read
// error.
func streamAudio(
	ctx context.Context,
	dispatcher *client.Dispatcher,
	requestId string,
	audio io.Reader,
	chunkBytes int,
) error {
	buf := make([]byte, chunkBytes)
	for {
		select {
		case <-ctx.Done():
			_ = sendEnd(dispatcher, requestId, true)
			return ctx.Err()
		default:
		}
		n, err := audio.Read(buf)
		if n > 0 {
			chunk := &memqlv1.MemqlClientMessage{
				Payload: &memqlv1.MemqlClientMessage_AiTranscribeStreamChunk{
					AiTranscribeStreamChunk: &memqlv1.AiTranscribeStreamChunk{
						RequestId: requestId,
						// Copy the slice -- buf is reused next iteration.
						Audio: append([]byte(nil), buf[:n]...),
					},
				},
			}
			if _, sendErr := dispatcher.Send(chunk); sendErr != nil {
				return fmt.Errorf("send Chunk: %w", sendErr)
			}
		}
		if err == io.EOF {
			return sendEnd(dispatcher, requestId, false)
		}
		if err != nil {
			_ = sendEnd(dispatcher, requestId, true)
			return err
		}
	}
}

func sendEnd(dispatcher *client.Dispatcher, requestId string, cancel bool) error {
	end := &memqlv1.MemqlClientMessage{
		Payload: &memqlv1.MemqlClientMessage_AiTranscribeStreamEnd{
			AiTranscribeStreamEnd: &memqlv1.AiTranscribeStreamEnd{
				RequestId: requestId,
				Cancel:    cancel,
			},
		},
	}
	if _, err := dispatcher.Send(end); err != nil {
		return fmt.Errorf("send End: %w", err)
	}
	return nil
}
